Issue2603 stuck on payout block

hewison-chris opened the pull request:
Fixing an issue with unknown txs not being fetched from other nodes when too many are requested.

hewison-chris committed:
Trivial: Prevent crash with block interval less than 5 secs

When the block rewards were calculted an interval of less than 5 seconds
would cause a division by zero.

hewison-chris committed:
Trivial: Align logging of rejected and accepted

This makes it easier to read the rejected tx logging by using the
prettify format.

omerfirmak commented on source/agora/test/LocalTransactions.d:
```diff
@@ -72,8 +73,20 @@ unittest
+        debug { import std.stdio : writefln; try { writefln!"txs = %s"(block.txs.prettify); } catch (Exception) {} }
+        return spent_txs += block.txs.length;
+    }
+    retryFor(all_txs_spent() == 6, 5.seconds);
```
All this can be replaced by `expectTxExternalization`, right? Although it might need an overload.

hewison-chris replied:
Yes you are right.

omerfirmak commented on source/agora/consensus/Ledger.d:
```diff
@@ -1473,7 +1473,14 @@ public class ValidatingLedger : Ledger
-        data.tx_set = this.getCandidateTransactions(next_height, max_txs);
+        if (clock.networkTime >= nomination_start_time + max(1, (this.params.BlockInterval.total!"seconds" / 2)))
+        {
+            data.tx_set = []; // Empty block to prevent blockchain getting stuck
```
How does this prevent us from getting stuck on the next height?

hewison-chris replied:
To be honest I am not sure yet. 
Maybe it will not but at least we will get blocks at the correct interval.

hewison-chris replied:
In fact it has always been getting stuck on payout block so next block will not have Coinbase tx and may be ok.

hewison-chris replied:
I am now trying a different approach by updating `combineCandidates`. Maybe this will fix it.

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1240,12 +1241,25 @@ extern(D):
+                        log.warn("combineCandidates: tx {} not in local pool so will skip this candidate",
+                            tx_hash);
+                        skip_candidate = true; // We will not include this candidate
+                        continue;
```
I think we can just `break` here

hewison-chris replied:
Yes I wasn't sure if `break` was better here or not. It is the last in the set so `continue` and `break` result in the same runtime behavior.

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1257,7 +1271,8 @@ extern(D):
                     total_rate = total_rate,
                 };
-                candidate_holders ~= candidate_holder;
+                if (!skip_candidate)
```
This still does not solve the root cause here, nodes still can select different candidates and that is probably perfectly normal.

Before they could select different ones because they calculate the fees different, now they can still do the same because they have different TXs in their pool. I think it was less likely to happen before, because a single TX that we don't include in fee calculations would have less of an impact.

We already have a TX catchup process anyways, missing TXs should not delay block creation indefinitely.

hewison-chris replied:
<img width="1808" alt="Screen Shot 2021-11-01 at 16 22 39" src="https://user-images.githubusercontent.com/11815575/139636226-931c0a5d-c746-40d5-9920-d22f1c424a07.png">

hewison-chris replied:
^ This is from the nodes in Production that are not slashed. This pattern has repeated for over a week now.

hewison-chris replied:
It looks like v2 has a tx which the others are not able to get during catchup.

hewison-chris committed:
Fix fetching the unknown txs

Prevent the call to getTransactions failing because of too many
unknown txs are being requested in a single call. Now each call has up
to 8 tx hashes to fetch.

hewison-chris committed:
Trivial: trace log for `FullNode.getTransactions`

hewison-chris committed:
Trivial warn level log line for unknown txs

omerfirmak commented on source/agora/network/Manager.d:
```diff
@@ -994,25 +995,42 @@ public class NetworkManager
-            foreach (tx; peer.client.getTransactions(unknown_txs))
+            // Fetch transactions in chunks
+            auto hashes = unknown_txs[].map!((Hash h) => h).array; // can not be lazy so use array
+            auto added = hashes.chunks(8)
```
Why did we have to limit the TXs per request, were we hitting a timeout or something?

hewison-chris replied:
When I reproduced it locally I found that limiting the number in the Set to 8 enabled the call via vibe.d to actually happen.

omerfirmak replied:
Sounds like a workaround to me but it might not be worth to dive deep into vibe.d for this.

hewison-chris replied:
The long term goal is to switch to RPC and use ranges so not worth the time at the moment.

omerfirmak commented on source/agora/network/Manager.d:
```diff
@@ -994,25 +995,42 @@ public class NetworkManager
-                try
-                {
-                    ledger.acceptTransaction(tx);
-                    unknown_txs.remove(tx.hashFull());
```
This is still needed, so that we don't fetch same TXs from different peers over and over again.

hewison-chris replied:
It is removed in the `ledger.acceptTransaction`.

omerfirmak replied:
Yeap, but that does not affect our local `unknown_txs` here. The for loop in `getUnknownTXs` iterating over the peers would request same TXs even if we got it from one of the previous peers.

hewison-chris replied:
Well spotted. I'll update.

hewison-chris replied:
Now it refreshes the set after each peer.
